src/OffsetArrays.jl

```diff
@@ -508,21 +508,22 @@ for OR in [:IIUR, :IdOffsetRange]
         @eval @inline function Base.getindex(r::$R, s::$OR)
             @boundscheck checkbounds(r, s)
             @inbounds pr = r[UnitRange(s)]
-            _maybewrapoffset(pr, axes(s,1))
+            _indexedby(pr, axes(s))
         end
     end
 
     # this method is needed for ambiguity resolution
     @eval @inline function Base.getindex(r::StepRangeLen{T,<:Base.TwicePrecision,<:Base.TwicePrecision}, s::$OR) where T
         @boundscheck checkbounds(r, s)
         @inbounds pr = r[UnitRange(s)]
-        _maybewrapoffset(pr, axes(s,1))
+        _indexedby(pr, axes(s))
     end
 end
 
 # eltype conversion
 # This may use specialized map methods for the parent
 Base.map(::Type{T}, O::OffsetArray) where {T} = parent_call(x -> map(T, x), O)
+Base.map(::Type{T}, r::Union{IIUR, IdOffsetRange}) where {T<:Real} = _indexedby(map(T, UnitRange(r)), axes(r))
```

Comment:
Just to confirm if I understand it correctly, this is the only change, right? All other changes seem to be clean-up work.

Reply:
The changes are in https://github.com/JuliaArrays/OffsetArrays.jl/pull/232/commits/8243635442a6278a5ce921d2858e6315e5a3c5d4, sorry about noise introduced by the cleanup

Reply:
> This is indeed type-piracy, but given that `IdentityUnitRange` is defined to be `<: AbstractUnitRange{Int}`, there is no real way to evaluate this in Base (basically there is no range type that has offset indices and non-Int eltype).

For forward compatibility in future, maybe:

```suggestion
Base.map(::Type{T}, r::IdOffsetRange) where {T<:Real} = _indexedby(map(T, UnitRange(r)), axes(r))
if eltype(IIUR) === Int
    # this is type piracy, but unless `IdentityUnitRange` becomes an abstract type, there is no way to support
    # eltype map in Base
    Base.map(::Type{T}, r::IIUR) where {T<:Real} = _indexedby(map(T, UnitRange(r)), axes(r))
end
```

Reply:
Something like this makes sense. Perhaps we may special-case the `Int` conversion as `map(::Type{Int}, r::IdentityUnitRange) = r`

Reply:
Specialization for Int should really live in Base...

Reply:
In that case perhaps the current implementation is sufficient, and we may add methods to base for `Int`

Reply:
The `Int` case is now added as a PR to `Base`: https://github.com/JuliaLang/julia/pull/40746

Reply:
> I think `eltpye(IIUR)` will always be Int?

Not if you extend its definition https://github.com/JuliaLang/julia/pull/39998#discussion_r594587450. That's why I say "forward compatibility" :)